pkg/cloud/services/networking/port.go

```diff
@@ -257,3 +258,37 @@ func (s *Service) GarbageCollectErrorInstancesPort(eventObject runtime.Object, i
 
 	return nil
 }
+
+// CollectPortSecurityGroups collects distinct securityGroups from port.SecurityGroups and port.SecurityGroupFilter fields.
+func (s *Service) CollectPortSecurityGroups(eventObject runtime.Object, portSecurityGroups *[]string, portSecurityGroupFilters []infrav1.SecurityGroupParam) (*[]string, error) {
+	var allSecurityGroupIDs []string
+	// security groups provided with the portSecurityGroupFilters fields
+	securityGroupFiltersByID, err := s.GetSecurityGroups(portSecurityGroupFilters)
+	if err != nil {
+		return portSecurityGroups, fmt.Errorf("error getting security groups: %v", err)
+	}
+	allSecurityGroupIDs = append(allSecurityGroupIDs, securityGroupFiltersByID...)
+	securityGroupCount := len(allSecurityGroupIDs)
+	// security groups provided with the portSecurityGroups fields
+	if portSecurityGroups != nil {
+		allSecurityGroupIDs = append(allSecurityGroupIDs, *portSecurityGroups...)
+	}
+	// generate unique values
+	uids := make(map[string]int)
+	for _, sg := range allSecurityGroupIDs {
+		if sg == "" {
+			continue
+		}
+		uids[sg] = 1
+		securityGroupCount++
```

Comment:
`securityGroupCount` is initialized to the number of SGs returned from the filter (line 271). Here you increment it by one for every single security group returned by either the filters, or the explicit id list. In effect, you are double counting the security groups from the filter.

This can be fixed by initializing `securityGroupCount := 0` in line 271.

This does not introduce any impacting bugs, only the slice for `distinctSecurityGroupIDs` will have a backing array that is too large.

Reply:
True, but even with `securityGroupCount := 0` at 271, we will still have extra space consumed. 
The reason is `allSecurityGroupIDs` could still contain duplicates. 

I will add an extra loop to count distinct values.

Reply:
Added `securityGroupCount := 0 `at 271, change the incremental to conditional, thanks.